library/control/test/workflow_manager_test.rb

```diff
@@ -386,22 +386,59 @@
       subject.control_file(repo_id)
     end
 
-    it "returns nil if the extracted package does not contain installation.xml" do
-      expect(File).to receive(:exist?).with(/installation\.xml\z/).and_return(false)
-      expect(subject.control_file(repo_id)).to be nil
-    end
+    context "if downloading and extracting worked" do
+      before do
+        allow(Dir).to receive(:glob).with(/installation-products/).and_return product_files
+        allow(Dir).to receive(:glob).with(/system-roles/).and_return role_files
 
-    it "returns the installation.xml path if the extracted package contains it" do
-      expect(File).to receive(:exist?).with(/installation.xml\z/).and_return(true)
-      # the returned path contains "/installation.xml" at the end
-      expect(subject.control_file(repo_id)).to end_with("/installation.xml")
-    end
+        allow(File).to receive(:exist?) do |name|
+          product_files.include?(name) || role_files.include?(name)
+        end
+
+        allow(File).to receive(:exist?).with(/installation\.xml\z/).and_return installation_xml
+      end
+
+      context "if the package contains a control file in the system-roles dir" do
+        let(:product_files) { [] }
+        let(:role_files) { ["/tmp/usr/share/system-roles/superyast.xml"] }
+        let(:installation_xml) { false }
+
+        it "returns the path of the control file" do
+          expect(subject.control_file(repo_id)).to eq "/tmp/usr/share/system-roles/superyast.xml"
+        end
+      end
+
+      context "if the package contains a control file in the installation-products dir" do
```

Comment:
maybe just for documentation purpose, can you write spec that if rpm contains combination of all methods, that installation-product has precedence?

Reply:
That is what I meant in https://github.com/yast/yast-yast2/pull/874#discussion_r240717962

Reply:
@jreidinger I was tempted to write such spec... and more. I didn't do it because I wanted to keep the simplicity of the current test suite (which doesn't test/document what happens when we have `/installation.xml` and also a role file or what happens when we have several role files) since you usually complain that I tend to over-complicate tests. 😏

But I will take your request as a green light to write one of my comprehensive tests. Be prepared. :wink: